File: client-go/client/utils/tablet.go
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package utils

import (
	"bytes"
	"encoding/binary"
	log "github.com/sirupsen/logrus"
	"sort"
)

type Tablet struct {
	DeviceId     string
	Measurements []string
	Values       []interface{}
	Timestamps   []int64
	Types        []int32
}

func (t *Tablet) GetRowNumber() int32 {
	return (int32)(len(t.Timestamps))
}

func (t *Tablet) GetDeviceId() string {
	return t.DeviceId
}

func (t *Tablet) GetMeasurements() []string {
	return t.Measurements
}

func (t *Tablet) GetBinaryTimestamps() []byte {
	buf := bytes.NewBuffer([]byte{})
	for i := 0; i < len(t.Timestamps); i++ {
		binary.Write(buf, binary.BigEndian, t.Timestamps[i])
	}
	return buf.Bytes()
}

func (t *Tablet) GetBinaryValues() []byte {
	buf := bytes.NewBuffer([]byte{})
	for i := 0; i < len(t.Types); i++ {
		switch t.Types[i] {
		case 0:
			binary.Write(buf, binary.BigEndian, t.Values[i].([]bool))
		case 1:
			tmp := t.Values[i].([]int32)
			binary.Write(buf, binary.BigEndian, &tmp)
		case 2:
			tmp := t.Values[i].([]int64)
			binary.Write(buf, binary.BigEndian, &tmp)
		case 3:
			tmp := t.Values[i].([]float32)
			binary.Write(buf, binary.BigEndian, &tmp)
		case 4:
			tmp := t.Values[i].([]float64)
			binary.Write(buf, binary.BigEndian, &tmp)
		case 5:
			values := t.Values[i].([]string)
			for index := range values {
				tmp := (int32)(len(values[index]))
				binary.Write(buf, binary.BigEndian, &tmp)
				buf.WriteString(values[index])
			}

		}
	}
	return buf.Bytes()
}

func (t *Tablet) GetTypes() []int32 {
	return t.Types
}

func (t *Tablet) SortTablet() {
	var timeIndexs = make(map[int64]int, t.GetRowNumber())
	for index := range t.Timestamps {
		timeIndexs[t.Timestamps[index]] = index
	}
	var keys []int64
	for timeValue := range timeIndexs {
		keys = append(keys, timeValue)
	}
	sort.Slice(keys, func(i, j int) bool { return keys[i] < keys[j] })
	t.Timestamps = keys
	for index := range t.Values {
		sortValue := sortList(t.Values[index], t.Types[index], timeIndexs, t.Timestamps)
		if sortValue != nil {
			t.Values[index] = sortValue
		} else {
			log.Error("unsupported data type ", t.Types[index])
		}
	}
}

func sortList(valueList interface{}, dataType int32, timeIndexs map[int64]int, timeStamps []int64) interface{} {
	switch dataType {
	case 0:
		boolValues := valueList.([]bool)
		sortedValues := make([]bool, len(boolValues))
		for index := range sortedValues {
			sortedValues[index] = boolValues[timeIndexs[timeStamps[index]]]
		}
		return sortedValues
	case 1:
		intValues := valueList.([]int32)
		sortedValues := make([]int32, len(intValues))
		for index := range sortedValues {
			sortedValues[index] = intValues[timeIndexs[timeStamps[index]]]
		}
		return sortedValues
	case 2:
		longValues := valueList.([]int64)
		sortedValues := make([]int64, len(longValues))
		for index := range sortedValues {
			sortedValues[index] = longValues[timeIndexs[timeStamps[index]]]
		}
		return sortedValues
	case 3:
		floatValues := valueList.([]float32)
		sortedValues := make([]float32, len(floatValues))
		for index := range sortedValues {
			sortedValues[index] = floatValues[timeIndexs[timeStamps[index]]]
		}
		return sortedValues
	case 4:
		doubleValues := valueList.([]float64)
		sortedValues := make([]float64, len(doubleValues))
		for index := range sortedValues {
			sortedValues[index] = doubleValues[timeIndexs[timeStamps[index]]]
		}
		return sortedValues
	case 5:
		stringValues := valueList.([]string)
		sortedValues := make([]string, len(stringValues))
		for index := range sortedValues {
			sortedValues[index] = stringValues[timeIndexs[timeStamps[index]]]
		}
		return sortedValues
	}
	return nil
}

File: client-py/iotdb/utils/Tablet.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#

import struct

from iotdb.utils.IoTDBConstants import TSDataType
from iotdb.utils.BitMap import BitMap


class Tablet(object):
    def __init__(
        self, device_id, measurements, data_types, values, timestamps, use_new=False
    ):
        """
        creating a tablet for insertion
          for example, considering device: root.sg1.d1
            timestamps,     m1,    m2,     m3
                     1,  125.3,  True,  text1
                     2,  111.6, False,  text2
                     3,  688.6,  True,  text3
        Notice: From 0.13.0, the tablet can contain empty cell
                The tablet will be sorted at the initialization by timestamps

        :param device_id: String, IoTDB time series path to device layer (without sensor).
        :param measurements: List, sensors.
        :param data_types: TSDataType List, specify value types for sensors.
        :param values: 2-D List, the values of each row should be the outer list element.
        :param timestamps: List.
        """
        if not use_new and len(timestamps) != len(values):
            raise RuntimeError(
                "Input error! len(timestamps) does not equal to len(values)!"
            )

        if not Tablet.check_sorted(timestamps):
            sorted_zipped = sorted(zip(timestamps, values))
            result = zip(*sorted_zipped)
            self.__timestamps, self.__values = [list(x) for x in result]
        else:
            self.__values = values
            self.__timestamps = timestamps

        self.__device_id = device_id
        self.__measurements = measurements
        self.__data_types = data_types
        self.__row_number = len(timestamps)
        self.__column_number = len(measurements)
        self.__use_new = use_new

    @staticmethod
    def check_sorted(timestamps):
        for i in range(1, len(timestamps)):
            if timestamps[i] < timestamps[i - 1]:
                return False
        return True

    def get_measurements(self):
        return self.__measurements

    def get_data_types(self):
        return self.__data_types

    def get_row_number(self):
        return self.__row_number

    def get_device_id(self):
        return self.__device_id

    def get_binary_timestamps(self):
        if not self.__use_new:
            format_str_list = [">"]
            values_tobe_packed = []
            for timestamp in self.__timestamps:
                format_str_list.append("q")
                values_tobe_packed.append(timestamp)

            format_str = "".join(format_str_list)
            return struct.pack(format_str, *values_tobe_packed)
        else:
            return self.__timestamps.tobytes()

    def get_binary_values(self):
        if not self.__use_new:
            format_str_list = [">"]
            values_tobe_packed = []
            bitmaps = []
            has_none = False
            for i in range(self.__column_number):
                bitmap = None
                bitmaps.insert(i, bitmap)
                if self.__data_types[i] == TSDataType.BOOLEAN:
                    format_str_list.append(str(self.__row_number))
                    format_str_list.append("?")
                    for j in range(self.__row_number):
                        if self.__values[j][i] is not None:
                            values_tobe_packed.append(self.__values[j][i])
                        else:
                            values_tobe_packed.append(False)
                            self.__mark_none_value(bitmaps, bitmap, i, j)
                            has_none = True

                elif self.__data_types[i] == TSDataType.INT32:
                    format_str_list.append(str(self.__row_number))
                    format_str_list.append("i")
                    for j in range(self.__row_number):
                        if self.__values[j][i] is not None:
                            values_tobe_packed.append(self.__values[j][i])
                        else:
                            values_tobe_packed.append(0)
                            self.__mark_none_value(bitmaps, bitmap, i, j)
                            has_none = True

                elif self.__data_types[i] == TSDataType.INT64:
                    format_str_list.append(str(self.__row_number))
                    format_str_list.append("q")
                    for j in range(self.__row_number):
                        if self.__values[j][i] is not None:
                            values_tobe_packed.append(self.__values[j][i])
                        else:
                            values_tobe_packed.append(0)
                            self.__mark_none_value(bitmaps, bitmap, i, j)
                            has_none = True

                elif self.__data_types[i] == TSDataType.FLOAT:
                    format_str_list.append(str(self.__row_number))
                    format_str_list.append("f")
                    for j in range(self.__row_number):
                        if self.__values[j][i] is not None:
                            values_tobe_packed.append(self.__values[j][i])
                        else:
                            values_tobe_packed.append(0)
                            self.__mark_none_value(bitmaps, bitmap, i, j)
                            has_none = True

                elif self.__data_types[i] == TSDataType.DOUBLE:
                    format_str_list.append(str(self.__row_number))
                    format_str_list.append("d")
                    for j in range(self.__row_number):
                        if self.__values[j][i] is not None:
                            values_tobe_packed.append(self.__values[j][i])
                        else:
                            values_tobe_packed.append(0)
                            self.__mark_none_value(bitmaps, bitmap, i, j)
                            has_none = True

                elif self.__data_types[i] == TSDataType.TEXT:
                    for j in range(self.__row_number):
                        if self.__values[j][i] is not None:
                            value_bytes = bytes(self.__values[j][i], "utf-8")
                            format_str_list.append("i")
                            format_str_list.append(str(len(value_bytes)))
                            format_str_list.append("s")
                            values_tobe_packed.append(len(value_bytes))
                            values_tobe_packed.append(value_bytes)
                        else:
                            value_bytes = bytes("", "utf-8")
                            format_str_list.append("i")
                            format_str_list.append(str(len(value_bytes)))
                            format_str_list.append("s")
                            values_tobe_packed.append(len(value_bytes))
                            values_tobe_packed.append(value_bytes)
                            self.__mark_none_value(bitmaps, bitmap, i, j)
                            has_none = True

                else:
                    raise RuntimeError(
                        "Unsupported data type:" + str(self.__data_types[i])
                    )

            if has_none:
                for i in range(self.__column_number):
                    format_str_list.append("?")
                    if bitmaps[i] is None:
                        values_tobe_packed.append(False)
                    else:
                        values_tobe_packed.append(True)
                        format_str_list.append(str(self.__row_number // 8 + 1))
                        format_str_list.append("c")
                        for j in range(self.__row_number // 8 + 1):
                            values_tobe_packed.append(bytes([bitmaps[i].bits[j]]))
            format_str = "".join(format_str_list)
            return struct.pack(format_str, *values_tobe_packed)
        else:
            bs_len = 0
            bs_list = []
            for i, value in enumerate(self.__values):
                if self.__data_types[i] == TSDataType.TEXT:
                    format_str_list = [">"]
                    values_tobe_packed = []
                    for str_list in value:
                        # Fot TEXT, it's same as the original solution
                        value_bytes = bytes(str_list, "utf-8")
                        format_str_list.append("i")
                        format_str_list.append(str(len(value_bytes)))
                        format_str_list.append("s")
                        values_tobe_packed.append(len(value_bytes))
                        values_tobe_packed.append(value_bytes)
                    format_str = "".join(format_str_list)
                    bs = struct.pack(format_str, *values_tobe_packed)
                else:
                    bs = value.tobytes()
                bs_list.append(bs)
                bs_len += len(bs)
            ret = memoryview(bytearray(bs_len))
            offset = 0
            for bs in bs_list:
                _l = len(bs)
                ret[offset : offset + _l] = bs
                offset += _l
            return ret

    def __mark_none_value(self, bitmaps, bitmap, column, row):
        if bitmap is None:
            bitmap = BitMap(self.__row_number)
            bitmaps.insert(column, bitmap)
        bitmap.mark(row)
